src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.cpp

```diff
@@ -0,0 +1,203 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.hpp"
+
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+//#include "Parallel/Printf.hpp"
+
+namespace {
+// This auxiliary function codes Eq. (B.34) of
+// F.Foucart's thesis, which is needed to find
+// the maximum value of S^2 allowed for the conservative
+// variables to be physical.
+double gW_MHD(const double W, const double B2sc, const double tsc,
+              const double mu) {
+  return (W + B2sc - tsc - 1.) * (W * W + B2sc * mu * mu * (B2sc + 2. * W)) -
+         B2sc / 2. -
+         B2sc * mu * mu / 2. * (W * W - 1. + 2. * W * B2sc + B2sc * B2sc);
+}
+}  // namespace
+
+namespace VariableFixing {
+FixConservatives::FixConservatives(
+    const double minimum_rest_mass_density_times_lorentz_factor) noexcept
+    : minimum_rest_mass_density_times_lorentz_factor_(
+          minimum_rest_mass_density_times_lorentz_factor) {}
+
+// clang-tidy: google-runtime-references
+void FixConservatives::pup(PUP::er& p) noexcept {  // NOLINT
+  p | minimum_rest_mass_density_times_lorentz_factor_;
+}
+
+void FixConservatives::operator()(
+    const gsl::not_null<Scalar<DataVector>*> tilde_d,
+    const gsl::not_null<Scalar<DataVector>*> tilde_tau,
+    const gsl::not_null<tnsr::i<DataVector, 3, Frame::Inertial>*> tilde_s,
+    const gsl::not_null<tnsr::I<DataVector, 3, Frame::Inertial>*> tilde_b,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric,
+    const Scalar<DataVector>& sqrt_det_spatial_metric) const noexcept {
+  // For the other limits, we have to define
+  //  \hat{B} = B*g^{1/4}/tildeD^{1/2}
+  //          = tildeB/g^{1/4}/tildeD^{1/2}
+  //  \hat{\tau} = tildeTau/tildeD
+  //  \hat{S} = tildeS/tildeD
+  // WE ALSO ASSUME THAT THE MINIMUM ENTHALPY IS h=1!!
+  // This can probably be modified.
+  //
+  // Then, two physical limits can be imposed
+  // (see Appendix B of F. Foucart's thesis)
+  // 1) (\hat{B})^2 < 2 \hat{\tau}
+  // 2) (\hat{S})^2 < SSmax
+  // with SSmax a complex function of the conservative
```

Comment:
@ffoucart I'm reading `complex` as real + imag, is that correct? If not maybe use `complicated`?